align values on pointer-size values to avoid false positives that ignore padding

The current code just performs a packing based off reducing the memories, without accounting for the fact that all struct fields have to be aligned, for example I just added structslop to the Go standard library in a private fork but it is reporting unrealistic sizes that are also odd; there is no way we'll have odd sizes, yet padding exists.

For an exhibit here is an error message
```shell
./mbitmap.go:132:15: struct{bytep *uint8; mask uint8; index uintptr}
has size 24, could be 17, rearrange to struct{bytep *uint8; index uintptr; 
mask uint8} for optimal size
```
the error says the value is 24, but could be 17; that's impossible; the size will be 24 on both 32-bit and 64-bit machines given that we pad.

I believe the remedy for this could be rounding up to the next multiple of the pointer size

--- a/fields_preserve_test.go
+++ b/fields_preserve_test.go
@@ -16,6 +16,7 @@ package structslop
 
 import (
 	"go/ast"
+	"go/build"
 	"go/importer"
 	"go/parser"
 	"go/token"
@@ -43,11 +44,17 @@ type s struct {
 		t.Fatal(err)
 	}
 
+	stdSizes := types.SizesFor(build.Default.Compiler, build.Default.GOARCH)
+	sizes := &sizes{
+		stdSizes: stdSizes,
+		maxAlign: stdSizes.Alignof(types.Unsafe.Scope().Lookup("Pointer").(*types.TypeName).Type()),
+	}
+
 	ast.Inspect(f, func(n ast.Node) bool {
 		if atyp, ok := n.(*ast.StructType); ok {
 			if tv, ok := info.Types[atyp]; ok {
 				styp := tv.Type.(*types.Struct)
-				optStruct := optimalStructArrangement(styp)
+				optStruct := optimalStructArrangement(sizes, styp)
 				if optStruct.Field(0) != styp.Field(0) {
 					t.Errorf("%v field order changed", styp.Field(0))
 				}

--- a/sizes.go
+++ b/sizes.go
@@ -1,0 +1,105 @@
+// Copyright 2020 Orijtech, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package structslop
+
+import (
+	"go/types"
+)
+
+type sizes struct {
+	stdSizes types.Sizes
+	maxAlign int64
+}
+
+func (s *sizes) Offsetsof(fields []*types.Var) []int64 {
+	offsets := make([]int64, len(fields))
+	var o int64
+	for i, f := range fields {
+		a := s.Alignof(f.Type())
+		o = align(o, a)
+		offsets[i] = o
+		o += s.Sizeof(f.Type())
+	}
+	return offsets
+}
+
+func (s *sizes) Sizeof(T types.Type) int64 {
+	switch t := T.Underlying().(type) {
+	case *types.Array:
+		return t.Len() * s.Sizeof(t.Elem())
+	case *types.Struct:
+		nf := t.NumFields()
+		if nf == 0 {
+			return 0
+		}
+		o := int64(0)
+		max := int64(1)
+		for i := 0; i < nf; i++ {
+			ft := t.Field(i).Type()
+			a, sz := s.Alignof(ft), s.Sizeof(ft)
+			if a > max {
+				max = a
+			}
+			if i == nf-1 && sz == 0 && o != 0 {
+				sz = 1
+			}
+			o = align(o, a) + sz
+		}
+		return align(o, max)
+	}
+	return s.stdSizes.Sizeof(T)
+}
+
+func (s *sizes) Alignof(T types.Type) int64 {
+	switch t := T.Underlying().(type) {
+	case *types.Array:
+		return s.Alignof(t.Elem())
+	case *types.Struct:
+		max := int64(1)
+		for i, nf := 0, t.NumFields(); i < nf; i++ {
+			if a := s.Alignof(t.Field(i).Type()); a > max {
+				max = a
+			}
+		}
+		return max
+	case *types.Slice, *types.Interface, *types.Basic:
+		return s.stdSizes.Alignof(T)
+	}
+
+	// All other types.
+	a := s.Sizeof(T)
+	if a < 1 {
+		return 1
+	}
+	// complex{64,128} are aligned like [2]float{32,64}.
+	if isComplex(T) {
+		a /= 2
+	}
+	if a > s.maxAlign {
+		return s.maxAlign
+	}
+	return a
+}
+
+// align returns the smallest x >= subject such that x % target == 0.
+func align(subject, target int64) int64 {
+	x := subject + target - 1
+	return x - x%target
+}
+
+func isComplex(typ types.Type) bool {
+	t, ok := typ.Underlying().(*types.Basic)
+	return ok && t.Info()&types.IsComplex != 0
+}

--- a/structslop.go
+++ b/structslop.go
@@ -37,6 +37,14 @@ var Analyzer = &analysis.Analyzer{
 }
 
 func run(pass *analysis.Pass) (interface{}, error) {
+	// Use custom sizes instance, which implements types.Sizes for calculating struct size.
+	// go/types and gc does not agree about the struct size.
+	// See https://github.com/golang/go/issues/14909#issuecomment-199936232
+	pass.TypesSizes = &sizes{
+		stdSizes: types.SizesFor(build.Default.Compiler, build.Default.GOARCH),
+		maxAlign: pass.TypesSizes.Alignof(types.Typ[types.UnsafePointer]),
+	}
+
 	inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
 
 	nodeFilter := []ast.Node{
@@ -53,7 +61,7 @@ func run(pass *analysis.Pass) (interface{}, error) {
 			return
 		}
 
-		r := checkSloppy(styp)
+		r := checkSloppy(pass, styp)
 		if !r.sloppy() {
 			return
 		}
@@ -72,8 +80,6 @@ func run(pass *analysis.Pass) (interface{}, error) {
 	return nil, nil
 }
 
-var sizes = types.SizesFor(build.Default.Compiler, build.Default.GOARCH)
-
 type result struct {
 	oldSize         int64
 	newSize         int64
@@ -84,16 +90,16 @@ func (r result) sloppy() bool {
 	return r.oldSize > r.newSize
 }
 
-func checkSloppy(origStruct *types.Struct) result {
-	optStruct := optimalStructArrangement(origStruct)
+func checkSloppy(pass *analysis.Pass, origStruct *types.Struct) result {
+	optStruct := optimalStructArrangement(pass.TypesSizes, origStruct)
 	return result{
-		oldSize:         sizes.Sizeof(origStruct),
-		newSize:         sizes.Sizeof(optStruct),
+		oldSize:         pass.TypesSizes.Sizeof(origStruct),
+		newSize:         pass.TypesSizes.Sizeof(optStruct),
 		suggestedStruct: optStruct,
 	}
 }
 
-func optimalStructArrangement(s *types.Struct) *types.Struct {
+func optimalStructArrangement(sizes types.Sizes, s *types.Struct) *types.Struct {
 	nf := s.NumFields()
 	fields := make([]*types.Var, nf)
 	for i := 0; i < nf; i++ {

--- a/testdata/src/struct/p.go
+++ b/testdata/src/struct/p.go
@@ -27,13 +27,13 @@ type s2 struct {
 	j int
 }
 
-type s3 struct { // want "struct{.+} has size 20, could be 16, rearrange to struct{y uint64; x uint32; z uint32} for optimal size"
+type s3 struct { // want "struct{.+} has size 24, could be 16, rearrange to struct{y uint64; x uint32; z uint32} for optimal size"
 	x uint32
 	y uint64
 	z uint32
 }
 
-type s4 struct { // want `struct{.+} has size 32, could be 20, rearrange to struct{_ \[0\]func\(\); i1 int; i2 int; a3 \[3\]bool; b bool} for optimal size`
+type s4 struct { // want `struct{.+} has size 40, could be 24, rearrange to struct{_ \[0\]func\(\); i1 int; i2 int; a3 \[3\]bool; b bool} for optimal size`
 	b  bool
 	i1 int
 	i2 int
@@ -41,14 +41,22 @@ type s4 struct { // want `struct{.+} has size 32, could be 20, rearrange to stru
 	_  [0]func()
 }
 
-type s5 struct { // want `struct{.+} has size 24, could be 20, rearrange to struct{y uint64; z \*httptest.Server; x uint32} for optimal size`
+type s5 struct { // want `struct{.+} has size 32, could be 24, rearrange to struct{y uint64; z \*httptest.Server; x uint32; t uint32} for optimal size`
 	x uint32
 	y uint64
 	z *httptest.Server
+	t uint32
 }
 
-type s6 struct { // want `struct{.+} has size 24, could be 20, rearrange to struct{y uint64; z \*s; x uint32} for optimal size`
+type s6 struct { // want `struct{.+} has size 32, could be 24, rearrange to struct{y uint64; z \*s; x uint32; t uint32} for optimal size`
 	x uint32
 	y uint64
 	z *s
+	t uint32
+}
+
+type s7 struct { // should be good, see #16
+	bytep *uint8
+	mask  uint8
+	index uintptr
 }

